File: tests/reboot/documentation/test_http.ts
import { Application, Reboot } from "@reboot-dev/reboot";
import * as http from "http";
import { strict as assert } from "node:assert";
import test from "node:test";
import { Greeter } from "../greeter_rbt.js";
import { GreeterServicer } from "../nodejs/greeter.js";

test("Tests for Reboot HTTP documentation", async (t) => {
  await t.test("Simple HTTP", async (t) => {
    const application = new Application({ servicers: [GreeterServicer] });

    application.http.get("/hello_world", async (context, req, res) => {
      res.json({ message: `Hello, world!` });
    });

    const rbt = new Reboot();
    await rbt.start();
    t.after(async () => {
      await rbt.stop();
    });

    await rbt.up(application, { localEnvoy: true });
    // Make sure we can access the HTTP endpoint more than once.
    for (let i = 0; i < 5; i++) {
      const data = await new Promise<{ message: string }>((resolve, reject) => {
        const req = http.request(
          new URL(rbt.url() + "/hello_world"),
          {
            method: "GET",
          },
          (res) => {
            let data = "";
            res.on("data", (chunk) => {
              data += chunk;
            });
            res.on("end", () => {
              resolve(JSON.parse(data) as { message: string });
            });
          }
        );

        req.on("error", (err) => {
          reject(err);
        });
        req.end();
      });
      assert(data.message == "Hello, world!", `${JSON.stringify(data)}`);
    }
  });

  await t.test("HTTP can use external context", async (t) => {
    const application = new Application({ servicers: [GreeterServicer] });

    application.http.post("/hello_greeter", async (context, req, res) => {
      const { title, name, adjective } = req.body as {
        title: string;
        name: string;
        adjective: string;
      };
      const [greeter] = await Greeter.create(context, {
        title,
        name,
        adjective,
      });

      const { message } = await greeter.greet(context, { name: "You" });

      res.json({ message });
    });

    const rbt = new Reboot();
    await rbt.start();
    t.after(async () => {
      await rbt.stop();
    });
    await rbt.up(application, { localEnvoy: true });

    // Make sure we can access the HTTP endpoint more than once.
    for (let i = 0; i < 5; i++) {
      const data = await new Promise<{ message: string }>((resolve, reject) => {
        const requestData = JSON.stringify({
          name: "Reboot",
          title: "Dr",
          adjective: "best doctor ever",
        });

        const req = http.request(
          new URL(rbt.url() + "/hello_greeter"),
          {
            method: "POST",
            headers: {
              "Content-Type": "application/json",
              "Content-Length": requestData.length,
            },
          },
          (res) => {
            let data = "";
            res.on("data", (chunk) => {
              data += chunk;
            });
            res.on("end", () => {
              resolve(JSON.parse(data) as { message: string });
            });
          }
        );

        req.on("error", (err) => {
          reject(err);
        });

        req.write(requestData);

        req.end();
      });

      assert(
        data.message == "Hi You, I am Dr Reboot the best doctor ever",
        `${JSON.stringify(data)}`
      );
    }
  });
});

File: tests/reboot/documentation/test_http.py
import aiohttp
import unittest
from fastapi import FastAPI, Request  # type: ignore[import]
from reboot.aio.applications import Application
from reboot.aio.external import ExternalContext
from reboot.aio.http import InjectExternalContext
from reboot.aio.tests import Reboot
from tests.reboot.greeter_rbt import Greeter
from tests.reboot.greeter_servicers import MyGreeterServicer

application = Application(servicers=[MyGreeterServicer])


@application.http.get("/hello_world")
def hello_world():
    return {"message": "Hello, world!"}


@application.http.post("/hello_greeter")
async def hello_greeter(
    request: Request,
    context: ExternalContext = InjectExternalContext,
):
    body = await request.json()
    greeter, _ = await Greeter.create(
        context,
        title=body['title'],
        name=body['name'],
        adjective=body['adjective'],
    )

    response = await greeter.greet(context, name="You")

    return {"message": response.message}


fastapi = FastAPI()


@fastapi.get("/{message}")
async def mount(message: str):
    return {"message": message}


def factory(_):
    return fastapi


application.http.mount("/mount", factory=factory)


class TestSomething(unittest.IsolatedAsyncioTestCase):

    async def asyncSetUp(self) -> None:
        self.rbt = Reboot()
        await self.rbt.start()

    async def asyncTearDown(self) -> None:
        await self.rbt.stop()

    async def test_http_new(self) -> None:

        await self.rbt.up(
            application,
            local_envoy=True,
        )

        # Make sure we can access the HTTP endpoint more than once.
        for _ in range(5):
            async with aiohttp.ClientSession() as session:
                async with session.request(
                    'GET',
                    self.rbt.url() + "/hello_world",
                ) as get:
                    self.assertEqual(get.status, 200)
                    self.assertEqual(
                        await get.json(), {"message": "Hello, world!"}
                    )

        # Make sure we can access the HTTP endpoint more than once.
        for _ in range(5):
            async with aiohttp.ClientSession() as session:
                async with session.request(
                    'POST',
                    self.rbt.url() + "/hello_greeter",
                    json={
                        "name": "Reboot",
                        "title": "Dr",
                        "adjective": "best doctor ever",
                    },
                ) as post:
                    self.assertEqual(post.status, 200)
                    self.assertEqual(
                        await post.json(),
                        {
                            "message":
                                "Hi You, I am Dr Reboot the best doctor ever",
                        },
                    )

        # Make sure we can also access the mount.
        async with aiohttp.ClientSession() as session:
            async with session.request(
                'GET',
                self.rbt.url() + "/mount/H3110",
            ) as get:
                self.assertEqual(get.status, 200)
                self.assertEqual(await get.json(), {"message": "H3110"})


if __name__ == '__main__':
    unittest.main()
